Kconfig

```diff
@@ -1904,7 +1904,7 @@ config HEAP_COLORATION
 
 config DEBUG_SYMBOLS
 	bool "Generate Debug Symbols"
-	default n
+	default y
```

Comment:
I think it could be useful for users that are willing to debug NuttX, but probably some users will forget to disable it and will ship the firmware with symbols enabled.

Reply:
but we normally ship firmware with binary or hex format.

Reply:
Enabling this option by default can lead to big f*ckups when ELF file with symbols is sent to users. 
DJI used to ship their pro drone cameras with debug symbols, which made it possible to reverse engineer their products but also showed a violation of the GPL license. I don't think the company was happy about that.
Potentially this is too dangerous for the fw provider.

Reply:
@raiden00pl @xiaoxiang781216 I did a test compiling stm32f103-minimum with and without debug symbols activated and in fact the final nuttx.bin is 99.999% equal in both cases. Actually the version with debug sysmbols enabled generated a binary 4 bytes smaller: 47464 vs 47468.

So, the only issue is fine the company ship the ELF binary to the user for some reason.

Reply:
@xiaoxiang781216 what about the "make debug" / "make release" approach?

Reply:
it need more work

Reply:
> it's tradeoff between the easy use and the more secure.

I would always opt for security over convenience.  The consequence of security problems is much worse than any invconvience since it could result in the exposure of intellectual property.,

You could make the default setting for CONFIG_DEBUG_SYMBOLS be "n" normally, but "y" if CONFIG_DEBUG_FEATURES is selected.  That default would probably be correct most of the time and, when it is not, it can still be overridden in the configuration.

A more hardline approach would be to make CONFIG_DEBUG_SYMBOLS  depend on CONFIG_DEBUG_FEATURES   That would probably not be wise.

In order for GDB to be usable, it needs not only symbols but usually also disabling optimization.  Should that have the same default setting?

Reply:
But symbols is also very useful in Postmortem debugging.

Reply:
> In order for GDB to be usable, it needs not only symbols but usually also disabling optimization. Should that have the same default setting?

Greg, that is not always true. Normally for ARM (ie. Cortex-M family) disabling optimization helps (or even is needed to show correctly functions parameters, etc), but for Xtensa (ESP32) disabling optimization doesn't help the debug and introduces some issues. It could be something specific for Xtensa GCC, not use. It was something that /me, @gustavonihei and @Ouss4 faced while debugging ESP32 chips.

Reply:
Please do not couple optimization setting and symbols. 

This is easier, but not the way the release code will really run. Chaining it may make an image not fit in Flash or change the target of a crash. 

Please do no make is easier for the user by turning on symbols automatically. 

It is a trivial change a user can make and does not expose them to unconscious security issues without their own actions.

Reply:
Choosing convenience over security is the worst possible option and is probably responsible for a large proportion of critical failures (infosec but also engineering in general). On the one hand, we have a small profit in the form of convenience for devs. On the other hand, we have the potential risk of a huge IP leak. From a risk management perspective, this change is just bad.